pkg/controller/database/redshift/controller.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package redshift
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsredshift "github.com/aws/aws-sdk-go-v2/service/redshift"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/password"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-aws/apis/database/v1alpha1"
+	"github.com/crossplane/provider-aws/pkg/clients/redshift"
+	"github.com/crossplane/provider-aws/pkg/controller/utils"
+)
+
+const (
+	errUnexpectedObject = "managed resource is not a Redshift custom resource"
+	errKubeUpdateFailed = "cannot update Redshift cluster custom resource"
+	errCreateFailed     = "cannot create Redshift cluster"
+	errModifyFailed     = "cannot modify Redshift cluster"
+	errDeleteFailed     = "cannot delete Redshift cluster"
+	errDescribeFailed   = "cannot describe Redshift cluster"
+	errUpToDateFailed   = "cannot check whether object is up-to-date"
+)
+
+// SetupRedshift adds a controller that reconciles Redshift clusters.
+func SetupRedshift(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.RedshiftGroupKind)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Redshift{}).
+		Complete(managed.NewReconciler(
+			mgr, resource.ManagedKind(v1alpha1.RedshiftGroupVersionKind),
+			managed.WithExternalConnecter(
+				&connector{kube: mgr.GetClient(),
+					newClientFn: redshift.NewClient,
+					awsConfigFn: utils.RetrieveAwsConfigFromProvider},
+			),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithConnectionPublishers(),
```

Comment:
Related to our meeting about the secret; an empty set of connection publishers are sent. That's why the secret is never published.

Reply:
The suggested change has been incorporated.